### ISSUE TITLE
FlashAttention implementation does not handle variable sequence lengths within a batch ? padding tokens participate in attention computation and distort softmax normalisation

### ISSUE THREAD

**Issue by eeshsaxena** [eeshsaxena/paper-implementations#4]
## Bug / Correctness Issue

The FlashAttention implementation processes batches of sequences. If sequences in a batch have different lengths, standard practice is to pad shorter sequences to the maximum length with a padding token. However, if the attention mask is not correctly applied inside the FlashAttention kernel, padding positions participate in the softmax computation.

**The softmax distortion problem:**
For a sequence of length L_real padded to L_max:
`
# Without masking (WRONG):
attn_scores = Q @ K.T  # shape: [L_max, L_max]
attn_weights = softmax(attn_scores)  # padding positions get non-zero weights
output = attn_weights @ V  # output at real positions is contaminated by padding values
`

The softmax over L_max positions gives non-zero attention weight to padding positions, so real token representations are influenced by padding embeddings. This is a **data leakage** from padding into real tokens.

**In standard attention this is fixed with:**
`python
attn_scores = attn_scores.masked_fill(padding_mask, float('-inf'))
attn_weights = softmax(attn_scores)  # padding positions ? weight 0
`

**FlashAttention-specific concern:**
The tiled computation in FlashAttention (computing attention in blocks to reduce memory) must apply the causal/padding mask within each block before the online softmax step. If masking is applied only post-hoc after tiling, the online softmax normalisation constant m and l are already contaminated.

**Questions:**
1. Does the implementation support attention masking (both causal and padding masks)?
2. Is the mask applied inside the tiled computation blocks, or only to the final output?
3. Has the implementation been tested on batches with variable-length sequences from a real dataset (e.g., WikiText-2 with standard padding) and compared against the HuggingFace reference implementation output?
